AMD SOF based generic SoundWire machine driver

vijendarmukunda opened the pull request:
This patch series is rebase version of PR#5034

vijendarmukunda committed:
ASoC: amd: acp: add soundwire machines for acp6.3 based platform

Add Soundwire machines for acp6.3 based platform.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: SOF: amd: add alternate machines for acp6.3 based platform

Add SoundWire machines as alternate machines for acp6.3 based platform.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: soc-acpi: add pci revision id field in mach params structure

Few IP's may have same PCI vendor id and device id and
different revision id. Add revision id field to the
'snd_soc_acpi_mach_params' so that using this field platform
specific implementation can be added in machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: SOF: amd: update mach params subsystem_rev variable

Add pci_rev variable in acp sof driver private data structure and assign
this value to mach_params structure subsystem_rev variable.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: amd/sdw_utils: add sof based soundwire generic machine driver

Add sof based Soundwire generic driver for amd platforms.
Currently support added for ACP6.3 based platforms.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw-utils: move soundwire machine driver soc ops

Move Intel SoundWire generic machine driver soc ops to common place
so that it can be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel: rename codec_info and dai_info structures names

To make it generic, rename structure 'sof_sdw_codec_info' as
'asoc_sdw_codec_info' and 'sof_sdw_dai_info' as 'asoc_sdw_dai_info'.
These structures will be moved to common header file so that it can
be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel: rename soundwire common header macros

Rename sof quirk macros, dai type and dai link macros with "SOC_SDW" tag.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel: rename soundwire codec helper functions

Rename SoundWire codec helper functions with "asoc_sdw" tag.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel: rename soundwire machine driver soc ops

Rename Soundwire generic machine driver soc ops with tag "asoc".

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel: rename maxim codec macros

Rename maxim codec part id macros.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel: rename ignore_pch_dmic variable name

Rename 'ignore_pch_dmic' variable name as 'ignore_internal_dmic'.
This variable will be moved to common header file and will be used by other
platform machine driver code.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel: move soundwire machine driver common structures

Move intel generic SoundWire machine driver common structures to
soc_sdw_utils.h file. These structures will be used in other platform
SoundWire machine driver code.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move soundwire machine driver helper functions

Move below Intel SoundWire machine driver helper functions to
soc_sdw_utils.c file so that it can be used by other platform machine
driver.
- asoc_sdw_is_unique_device()
- asoc_sdw_get_codec_name()

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move dmic codec helper function

Move generic dmic codec helper function implementation to
sdw_utils folder so that this function can be used by other platform
machine drivers.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rtk dmic helper functions

Move rtk SoundWire dmic helper functions implementation to sdw_utils
folder to make it generic.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rt712 sdca helper functions

Move RT712 SDCA codec helper file to sdw_utils folder so that these
helper functions can be used by other platform machine drivers.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rt722 sdca helper functions

Move RT722 SDCA codec helper file to sdw_utils folder to make it generic.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel: split soundwire machine driver private data

Split intel generic SoundWire machine driver private data into two
structures. One structure is generic one which can be used by other
platform machine driver and the other one is intel specific one.
Move generic machine driver private data to soc_sdw_utils.h.
Define a void pointer in generic machine driver private data structure
and assign the vendor specific structure in mc_probe() call.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rt5682 codec helper function

Move rt5682 sdw codec helper function to common place holder to make it
generic.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rtk jack common helper functions

Move RTK codec jack common helper functions to common place holder
(sdw_utils folder) to make it generic so that it will be used by
other platform machine driver code.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rt700 and rt711 codec helper functions

Move RT700 and RT711 Soundwire codec helper functions to common
place holder so that it can be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move rtk amp codec helper functions

Move RTK amp codec helper functions related implementation to common
place holder to make it generic so that these helper functions will be
used by other platform machine driver modules.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move cirrus soundwire codec helper functions

To make it generic, move Cirrus Soundwire codec helper functions to
common place holder so that it can be used by other platform machine
driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move maxim codec helper functions

Move maxim codec helper functions to common place holder so that
it can be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move dai id common macros

Move dai id common macros from intel SoundWire generic driver to
soc_sdw_utils.h file so that it can be used by other platform machine
driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move soundwire dai type macros

Move SoundWire dai type macros to common header file(soc_sdw_util.h).
So that these macros will be used by other platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move soundwire codec_info_list structure

SoundWire 'codec_info_list' structure is not a platform specific one.
Move codec_info_list structure to common file soc_sdw_utils.c.
Move codec helper functions which uses codec_info_list structure to common
place holder and rename the function by adding _sdw tag. This will allow
to use 'codec_info_list' structure and it's helper functions in other
platform machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: intel/sdw_utils: move machine driver dai link helper functions

Move machine driver dai link helper functions to common place holder,
So that it can be used by other platform machine driver.
Rename these functions with "asoc_sdw" tag as a prefix.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda committed:
ASoC: sdw_utils: refactor sof_sdw_card_late_probe function

Refactor sof_sdw_card_late_probe() function and derive a generic
function soc_sdw_card_late_probe() function which can be used by
SoundWire generic machine driver.

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

plbossart commented on sound/soc/sdw_utils/soc_sdw_rt712_sdca.c:
```diff
@@ -1,8 +1,10 @@
 // SPDX-License-Identifier: GPL-2.0-only
+// This file incorporates work covered by the following copyright notice:
 // Copyright (c) 2023 Intel Corporation
+// Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
please check your copyright, it sometimes says All rights reserved and sometimes it doesn't.

vijendarmukunda replied:
will fix it.

vijendarmukunda replied:
Fixed.

plbossart commented on sound/soc/sof/amd/Kconfig:
```diff
@@ -23,6 +23,7 @@ config SND_SOC_SOF_AMD_COMMON
 	select SND_AMD_ACP_CONFIG
 	select SND_SOC_SOF_XTENSA
 	select SND_SOC_SOF_ACP_PROBES
+	select SND_SOC_ACPI_AMD_MATCH
```
this is weird, you have multiple selections of SND_SOC_ACPI.

config SND_SOC_ACPI_AMD_MATCH
	tristate
	select SND_SOC_ACPI if ACPI

vijendarmukunda replied:
its similar to Intel Kconfig option 'config 'SND_SOC_ACPI_INTEL_MATCH'.

plbossart replied:
ok, I don't fully remember why we did this, but if it ain't broke don't fix it...

plbossart commented on sound/soc/amd/acp/soc_amd_sdw_common.h:
```diff
@@ -0,0 +1,47 @@
+#include <sound/soc.h>
+#include <sound/soc_sdw_utils.h>
+
+#define MAX_NO_PROPS		2
```
isn't this defined already?

vijendarmukunda replied:
Yes. This can be dropped.

vijendarmukunda replied:
Fixed.

plbossart reviewed (commented):
Looks pretty good to me @vijendarmukunda 

I only have minor comments, time for others to comment as well. @bardliao @charleskeepax on your mark, set, go!

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+	{
+		.callback = sof_sdw_quirk_cb,
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "AMD"),
```
that looks like a very broad filter, is this needed?

vijendarmukunda replied:
As mentioned earlier, currently we are testing with Codec evaluation boards. Once we get final platform , we are going to change this logic.

vijendarmukunda replied:
Fixed. currently updated DMI check to ACP63 platform.

plbossart commented on sound/soc/amd/acp/soc_amd_sdw_common.h:
```diff
@@ -0,0 +1,46 @@
+#define SW0_AUDIO2_RX	5
+
+#define SW1_AUDIO0_TX	0
+#define SW1_AUDIO0_RX	1
```
the prefixes are inconsistent, should be SDW_AMD or something.

vijendarmukunda replied:
As this header file itself , AMD specific, I don't think we need to follow prefix convention here

vijendarmukunda replied:
we will change the prefix convention with ACP as prefix, because these macros may vary based on ACP version.

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+	dai_links->codecs = codecs;
+	dai_links->num_codecs = codecs_num;
+	dai_links->dpcm_playback = playback;
+	dai_links->dpcm_capture = capture;
```
is this still needed with Morimoto-san changes? I think we removed this on the Intel side?

vijendarmukunda replied:
SOF base hasn't pulled Moroimoto-san changes yet. same case with sof_sdw.c file. Once the Morimoto-san's patch  series got merged in to SOF github , i can rebase those changes

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+	}
+
+	return 0;
+}
```
could this be moved to a helper?

vijendarmukunda replied:
As this function being used by only AMD code, that's why didn't moved this code to a helper.

charleskeepax replied:
It would really make sense to pull in similar refactor as I did on the Intel machine driver, the resultant code is quite a bit cleaner and supports several corner cases this older style did not.

vijendarmukunda replied:
@charleskeepax : This is the initial version code.  We are going to add support for legacy(NO DSP) solution soon. 
We are going to refactor code similar to Intel machine driver code and push the incremental patch series.

charleskeepax replied:
Fair enough, if its getting done fairly soon I don't have any great objection.

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+	} while (adr_link->mask);
+
+	return 0;
+}
```
could this be moved to a helper?

vijendarmukunda replied:
As this function being used by only AMD code, that's why didn't moved this code to a helper.

vijendarmukunda replied:
Moved to helper

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+	codec->dai_name = codec_info->dais[dai_index].dai_name;
+	dev_err(dev, "codec->dai_name:%s\n", codec->dai_name);
+	return 0;
+}
```
could this be moved to a helper?

vijendarmukunda replied:
As this function being used by only AMD code, that's why didn't moved this code to a helper.

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+		      &dlc[0], 1, &dlc[1], 1, init, ops);
+
+	return 0;
+}
```
could this be moved to a helper?

vijendarmukunda replied:
No , It can't be moved. because it invokes init_dai_link() function which is static one.

vijendarmukunda replied:
moved to common place holder

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+	dai_links->dpcm_capture = capture;
+	dai_links->init = init;
+	dai_links->ops = ops;
+}
```
could this be moved to a helper?

vijendarmukunda replied:
platform component differs for Intel and AMD. I think time being , we can keep this function as static.

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+		}
+	}
+	return 0;
+}
```
could this be moved to a helper?

vijendarmukunda replied:
As this function being used by AMD only, thats why didn't moved code as a common helper.

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,779 @@
+			"SDW%d-PIN%d-CAPTURE",
+			"SDW%d-PIN%d-PLAYBACK-%s",
+			"SDW%d-PIN%d-CAPTURE-%s",
+		};
```
this seems to be the only difference with Intel, can we refactor further?

vijendarmukunda replied:
The key difference between Intel & AMD are
1) Port's allocation and mapping for TX & RX on Manager side.
2) AMD uses static mapping as our port mapping is different for TX & RX for per ACP variant. This is going to change for upcoming platforms(new ACP variants)
Currently Intel implementation is differs for creating SDW DAI link creation logic.
We are going to revisit the code, as we are going to add support for No DSP solution(legacy solution) where we are going to implement common functions for legacy and sof stacks soon.
We can revisit this logic again.

vijendarmukunda replied:
We are pushing these changes as initial version as things getting delayed. We are going to refactor the code further in upcoming patch series.

plbossart reviewed (changes requested):
The last patch needs a bit more work IMHO, still too much duplication of complicated code that we don't want to fork.

The rest looks ok, thanks!

plbossart commented:
SOFCI TEST

bardliao commented on sound/soc/intel/boards/bridge_cs35l56.c:
```diff
@@ -127,7 +127,7 @@ int bridge_cs35l56_add_sidecar(struct snd_soc_card *card,
 int bridge_cs35l56_spk_init(struct snd_soc_card *card,
 			    struct snd_soc_dai_link *dai_links,
-			    struct sof_sdw_codec_info *info,
+			    struct asoc_sdw_codec_info *info,
```
Commit message: soc_sdw_codec_info -> asoc_sdw_codec_info and soc_sdw_dai_info -> asoc_sdw_dai_info as the code does.

vijendarmukunda replied:
Will update the commit message

vijendarmukunda commented:
> The last patch needs a bit more work IMHO, still too much duplication of complicated code that we don't want to fork.
> 
> The rest looks ok, thanks!

Agreed. We are going to refactor the code . We need to add support for newer ACP varaints.
We would like to implement as an incremental patch series.

bardliao reviewed (dismissed):
Overall looks good to me. The duplicated code can be removed later.

plbossart commented:
We can't merge this because it breaks CI, the new module needs to be added to sof-test/tools/kmod

https://sof-ci.01.org/linuxpr/PR5068/build3852/devicetest/index.html?model=CML_RVP_SDW-ipc3&testcase=check-sof-logger

rmmod: ERROR: Module soundwire_bus is in use by: snd_soc_sdw_utils

vijendarmukunda commented:
> We can't merge this because it breaks CI, the new module needs to be added to sof-test/tools/kmod
> 
> https://sof-ci.01.org/linuxpr/PR5068/build3852/devicetest/index.html?model=CML_RVP_SDW-ipc3&testcase=check-sof-logger
> 
> rmmod: ERROR: Module soundwire_bus is in use by: snd_soc_sdw_utils

Not sure how to push changes to sof-test.git along with this PR#5068. 
[0001-kmod-add-snd_soc_sdw_utils-module.patch](https://github.com/user-attachments/files/15971415/0001-kmod-add-snd_soc_sdw_utils-module.patch)

Looks like i don't have permissions for submitting changes to sof-test.git tree.
Attached patch for reference.

plbossart commented:
SOFCI TEST

plbossart commented:
The patch is wrong @vijendarmukunda the modules need to be removed in the reverse order.

RMMOD	snd_soc_sdw_utils
rmmod: ERROR: Module snd_soc_sdw_utils is in use by: snd_soc_sof_sdw

vijendarmukunda commented:
> The patch is wrong @vijendarmukunda the modules need to be removed in the reverse order.
> 
> RMMOD snd_soc_sdw_utils rmmod: ERROR: Module snd_soc_sdw_utils is in use by: snd_soc_sof_sdw



> The patch is wrong @vijendarmukunda the modules need to be removed in the reverse order.
> 
> RMMOD snd_soc_sdw_utils rmmod: ERROR: Module snd_soc_sdw_utils is in use by: snd_soc_sof_sdw

Yes you are correct. Sorry It's my bad. Attached updated patch.
[0001-kmod-add-snd_soc_sdw_utils-module.patch](https://github.com/user-attachments/files/15974409/0001-kmod-add-snd_soc_sdw_utils-module.patch)

vijendarmukunda committed:
ASoC: intel/sdw_utils: refactor init_dai_link() and init_simple_dai_link()

To make it generic, refactor existing implementation for
init_dai_link() and init_simple_dai_link() as mentioned below.
- Move init_dai_link() and init_simple_dai_link() to common place holder
- Rename the functions with "asoc_sdw" as prefix.
- Pass the platform specific 'platform_component' structure and its size as
arguments for init_simple_dai_link() function and allocate one more
extra dlc for platform component.
- Pass the 'platform_component' and 'num_platforms' as arguments for
init_dai_link().

Signed-off-by: Vijendar Mukunda <Vijendar.Mukunda@amd.com>

vijendarmukunda commented:
Moved most of the helper functions to common place holder. Only SoundWire dai creation logic needs to be refactored. As mentioned earlier, we will make the changes as an incremental patch series later. I think this is good enough to push as an initial version code.
@plbossart @bardliao : could you please review the latest changes?

plbossart commented:
Looks like we have major CI regressions with this PR @vijendarmukunda, it'll take time to figure out what's wrong.

vijendarmukunda commented:
> Looks like we have major CI regressions with this PR @vijendarmukunda, it'll take time to figure out what's wrong.

I think module remove script fix has to updated. 
https://github.com/thesofproject/sof-test/pull/1210 - this fix is not going to solve the module remove error.
could you please try with latest shared one?

vijendarmukunda commented:
> Looks like we have major CI regressions with this PR @vijendarmukunda, it'll take time to figure out what's wrong.

Okay will wait for further update.

bardliao commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -932,5 +932,59 @@ int asoc_sdw_card_late_probe(struct snd_soc_card *card)
+
+	asoc_sdw_init_dai_link(dev, dai_links, be_id, name, playback, capture,
+			       &dlc[0], 1, &dlc[1], num_platforms,
+			       &dlc[2], 2, init, ops);
```
Why change codecs_num to 2?

vijendarmukunda replied:
its typo error. will correct the index

bardliao commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -932,5 +932,59 @@ int asoc_sdw_card_late_probe(struct snd_soc_card *card)
+	dlc[1].name = platform_comp_name;
+
+	dlc[1].name = codec_name;
+	dlc[1].dai_name = codec_dai_name;
```
Should those be dlc[2]?

vijendarmukunda replied:
yes correct. will update it.

vijendarmukunda commented:
Posted changes after fixing wrong index update and codec count in asoc_sdw_init_simple_dai_link().

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -932,5 +932,59 @@ int asoc_sdw_card_late_probe(struct snd_soc_card *card)
+
+	/* Allocate three DLCs one for the CPU, one for platform and one for the CODEC */
+	dlc = devm_kcalloc(dev, 3, sizeof(*dlc), GFP_KERNEL);
+	if (!dlc || !name || !cpu_dai_name || !codec_name || !codec_dai_name)
```
shouldn't you test for !platform_comp_name to be consistent?

vijendarmukunda replied:
yes you are correct. it's a miss. will add it.

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -932,5 +932,59 @@ int asoc_sdw_card_late_probe(struct snd_soc_card *card)
+	dlc[1].name = platform_comp_name;
+
+	dlc[2].name = codec_name;
+	dlc[2].dai_name = codec_dai_name;
```
the commit message needs to be expanded, it's not just a move but also addition of functionality

vijendarmukunda replied:
Agreed. Will update the commit message.

plbossart commented on sound/soc/amd/acp/amd-acp63-acpi-match.c:
```diff
@@ -0,0 +1,89 @@
+};
+EXPORT_SYMBOL(snd_soc_acpi_amd_acp63_sof_sdw_machines);
+
+MODULE_LICENSE("GPL");
```
Missing MODULE_DESCRIPTION()

vijendarmukunda replied:
will fix it.

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,600 @@
+	return 0;
+}
+
+static const char * const type_strings[] = {"SimpleJack", "SmartAmp", "SmartMic"};
```
newline

vijendarmukunda replied:
will fix it.

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,600 @@
+		i++;
+	}
+
+	  /* find codec info to create BE DAI */
```
indentation

vijendarmukunda replied:
will fix it.

plbossart commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -986,5 +986,139 @@ int asoc_sdw_init_simple_dai_link(struct device *dev, struct snd_soc_dai_link *d
+	}
+	return 0;
+}
+EXPORT_SYMBOL_NS(asoc_sdw_get_dailink_info, SND_SOC_SDW_UTILS);
```
I don't get those two helpers, shouldn't they be used in the Intel case?

Also it's probably better to add these helpers in a separate patch to make the review easier.

vijendarmukunda replied:
@plbossart : as mentioned earlier, Intel's logic is different. These helper functions are being used by AMD only. As mentioned as this code being to be refactored, is it okay to keep the two functions as static in amd machine driver for time being as an initial version?

plbossart replied:
if the helper is not shared, then it should not have an asoc_sdw prefix and it should remain in the AMD stuff. Thanks!

vijendarmukunda replied:
Thanks for clarification. will make the changes.

plbossart reviewed (commented):
Still a couple of nit-picks.

I tested this myself on MTL and the results look ok, so one more round and we should be good to go.

vijendarmukunda commented:
Fixed nit-picks and respin the patch series.

vijendarmukunda commented:
> Still a couple of nit-picks.
> 
> I tested this myself on MTL and the results look ok, so one more round and we should be good to go.

Fixed nit-picks and respin the patch series

charleskeepax commented on sound/soc/sdw_utils/soc_sdw_utils.c:
```diff
@@ -0,0 +1,990 @@
+	},
+};
+EXPORT_SYMBOL_NS(codec_info_list, SND_SOC_SDW_UTILS);
+
```
Feels a bit weird having all this stuff at the ASoC level, is this not a specific machine driver implementation, one that is shared between AMD and Intel now, but its not really generic, it is data for a specific implementation. That said I am not sure exactly where it should live, maybe the sof level? Are we calling this the sof machine driver implementation?

vijendarmukunda replied:
@charleskeepax: Soon we are going to add Legacy (NO DSP) machine driver which will refer same codec_info_list structure and helpers pulled in sdw_utils folder will be utilized by our legacy machine driver.  The reason beyond pulling the common helpers/structures at one place , so that Intel & AMD generic machine driver code base going to refer these common structures and helper functions. This avoids code duplication for example  codec helper functions are not specific to platform.

plbossart replied:
what makes you think this is implementation-specific @charleskeepax?

This table augments the basic dai descriptors by adding SDCA-type qualifiers that help create dailinks based on an ACPI description. That ACPI description follows the DisCo spec, it's not machine or SOC-specific.

I think the concept maps well for all SDCA topologies, we have dedicated streaming IT/OT for different functionality, I don't see currently a risk of a DAI/terminal being used for machine-specific routing. If you want to access to jack output or an ultrasound render terminal you are required to rely on an SDCA-defined terminal.

charleskeepax replied:
Now to be clear I don't have a mega-strong objection here, if Mark is happy to merge it I don't really mind it going in as is and it is not like putting this in ASoC prevents someone doing something else either.

But the code is very clearly one particular way of implementing things. Even the "acpi" bits are not some standardised acpi thing, they are just a bunch of tables contained in the machine driver. Basically the information encoding what the DAI links look like is partitioned between the acpi structs and the codec_info structs, but there is no reason to believe every machine driver would always want to make those same choices of how to partition that information. Even more so when you get to things like the init and exit callbacks, which even more explicitly depend on a specific implementation.

The way it is laid out does likely map reasonably well to SDCA, as they acpi structs fill in mostly what is missing from the disco description, in that case most of the information in the codec_info side should not be required as at that point it is either coming from the definition of the topology for the function or from the disco description of it.

plbossart replied:
agree, this list will probably disappear at some point and will be replaced by properties read from DisCo. There's no intention to make the integration of every codec in this list a requirement, it's more of a stop-gap solution until we have a better one.

vijendarmukunda commented:
@bardliao @plbossart @charleskeepax : could you please review the changes?  I need to push a patch series which is dependent on current PR.

vijendarmukunda commented:
@bhiregou @bhiregoudar  @sunilkumardommati : could you please review the patch series?

vijendarmukunda commented:
@plbossart : If you don't have further objections, could you please approve this PR? We need to submit the other patch series which is dependent on current PR?

plbossart commented:
SOFCI TEST

plbossart commented:
SOFCI TEST

plbossart commented:
One last try with CI, everything should be back to normal now.

plbossart commented:
SOFCI TEST

plbossart commented:
all tests ok -> merging. I don't like doing this on a Friday afternoon but we've got to move on. If any issues are found we'll know it fairly soon with Intel daily tests.

vijendarmukunda commented:
@plbossart : Tried to rebase PR#5068 on top of for-next branch for sending patches to alsa mailing list.
Observing merge conflicts as few patches related to sof_sdw.c file changes need to be pushed first.
Could you please push the dependent patches first?

plbossart commented:
@vijendarmukunda we're not going to push the Dell SKUs until we have confirmation, the simplest solution is to revert the patches. I just did this in topic/sof-dev-rebase, which will also contain the latest broonie/for-next updates. I'll update this shortly with a different PR.

We have to wait two weeks before submitting this series upstream, the merge window is open and maintainers are busy maintaining stuff.

plbossart commented:
@vijendarmukunda please double-check the topic/sof-dev-rebase branch (sof-dev-rebase-20240715)

vijendarmukunda commented:
@plbossart : I have checked sof-dev-rebase-20240715 base. PR 5068 & PR 5103 changes are merged. Will wait till merging window is open.

plbossart commented on sound/soc/amd/acp/acp-sdw-sof-mach.c:
```diff
@@ -0,0 +1,737 @@
+	 * codecs on used platform
+	 */
+	for (i = 0; i < ctx->codec_info_list_count; i++)
+		amp_num += codec_info_list[i].amp_num;
```
@vijendarmukunda there's a new warning on this patch:

````
commit: 8245a1411fef45a208f182c176660107b7f582aa [32/45] ASoC: amd/sdw_utils: add sof based soundwire generic machine driver
config: x86_64-allyesconfig (https://download.01.org/0day-ci/archive/20240716/202407160447.9KCzYnRz-lkp@intel.com/config)
compiler: clang version 18.1.5 (https://github.com/llvm/llvm-project 617a15a9eac96088ae5e9134248d8236e34b91b1)
reproduce (this is a W=1 build): (https://download.01.org/0day-ci/archive/20240716/202407160447.9KCzYnRz-lkp@intel.com/reproduce)

If you fix the issue in a separate patch/commit (i.e. not just a new version of
the same patch/commit), kindly add following tags
| Reported-by: kernel test robot [<lkp@intel.com>](mailto:lkp@intel.com)
| Closes: https://lore.kernel.org/oe-kbuild-all/202407160447.9KCzYnRz-lkp@intel.com/

All warnings (new ones prefixed by >>):

>> sound/soc/amd/acp/acp-sdw-sof-mach.c:644:6: warning: variable 'amp_num' set but not used [-Wunused-but-set-variable]
     644 |         int amp_num = 0, i;
         |             ^
   1 warning generated.
````

Care to submit a fixup?

vijendarmukunda replied:
will submit the fix up.

vijendarmukunda replied:
@plbossart : posted fix : https://github.com/thesofproject/linux/pull/5107

vijendarmukunda commented:
@plbossart : Need your suggestion.  Should  I push  PR 5068 & PR 5103 at one go for alsa devel mailing list rather than separate patch series?

vijendarmukunda commented:
> On 7/31/24 09:17, vijendarmukunda wrote: @plbossart <https://github.com/plbossart> : Need your suggestion. Should I push PR 5068 & PR 5103 at one go for alsa devel mailing list rather than separate patch series for each PR?
> They should be two separate series, one is SoundWire related and touches the intel stuff, the other is AMD/SOF centric. Please make sure you take all the patches from topic/sof-dev-rebase, that way you get all the Reviewed-by tags for free.

Okay. We will push the PR#5068 as first one , pulling pull the patches from topi/sof-dev-rebase branch . Once it get merged we will send the PR5103 patch series on top of that.

vijendarmukunda commented:
@plbossart : I am able to push only 21 patches for alsa devel mailing list. I got session timeout. Do I need to resend the complete patch series again?
Please suggest, how can i send complete patch set ?

plbossart commented:
no idea, never seen this issue on my side with git send-email. 
Could be a problem on your IT side or the ALSA mail server maintained by @perexg

vijendarmukunda commented:
> no idea, never seen this issue on my side with git send-email. Could be a problem on your IT side or the ALSA mail server maintained by @perexg

I have increased smtp server mail timeout. Will send the patch series again once i get reply from Mark brown.

vijendarmukunda commented:
still no luck. sent a mail to @perexg and mark brown and also checking internally on AMD side.

perexg commented:
We can do hardly do something with the local outgoing SMTP server configuration. It's a first hop SMTP server issue.

vijendarmukunda commented:
> We can do hardly do something with the local outgoing SMTP server configuration. It's a first hop SMTP server issue.

Thanks Will check SMTP server configuration on our side.
